src/Compilers/Core/Portable/SourceGeneration/GeneratorDriver.cs

```diff
@@ -320,10 +325,28 @@ private static GeneratorState SetGeneratorException(CommonMessageProvider provid
 
         internal static string GetFilePathPrefixForGenerator(ISourceGenerator generator)
         {
-            var type = generator.GetType();
+            var type = GetGeneratorType(generator);
             return Path.Combine(type.Assembly.GetName().Name ?? string.Empty, type.FullName!);
         }
 
+        public static Type GetGeneratorType(Type generatorType)
```

Comment:
Given this is a public API please add a comment indicating what this method does. The behavior is not going to be obvious at a glance to the customer.

Reply:
I ended up moving this to `AnalyzerFileReference` given that is the place that actually does the wrapping, and you'll only need to unwrap the type if you loaded via this mechanism. Added comments and test too.